beetsplug/replaygain.py

```diff
@@ -277,6 +299,246 @@ def end_element_handler(name):
             out.append(album_gain["album"])
         return out
 
+    def use_ebu_r128(self):
+        """Set this Backend up to use EBU R128."""
+        self.method = '--ebu'
+
+
+# ffmpeg backend
+class FfmpegBackend(Backend):
+    """A replaygain backend using ffmpegs ebur128 filter.
+    """
+    def __init__(self, config, log):
+        super(FfmpegBackend, self).__init__(config, log)
+        config.add({
+            "peak": "true"
+        })
+        self._peak_method = config["peak"].as_str()
+        self._target_level = db_to_lufs(config['targetlevel'].as_number())
+        self._ffmpeg_path = "ffmpeg"
+
+        # check that ffmpeg is installed
+        # TODO: check if ffmpeg version is sufficient
```

Comment:
Should we really check the ffmpeg version here? If so, how? Most other backends do not do this.

I have not tested any versions < libavfilter 7.16.100, but we can most likely support [libavfilter 6.67.100](https://git.videolan.org/?p=ffmpeg.git;a=commit;h=005d058f4230f3207ebcf1131df7426d4f57392f).

Before that ffmpeg had to be configured with the `--enable-libebur128` option. Should we maybe just raise a warning for ffmpeg versions < 6.67.100? That would allow users with that feature enabled to still use this backend.

If we decide to check the version: Are there any utility functions relevant to this in beets or the python stdlib? How should the libavfilter version be parsed and compared?

Reply:
Good question. If this were super easy to do, I would probably say yes—it wouldn’t hurt to issue a reasonable warning if the backend is missing. But it actually seems pretty complicated to do. You’d have to get the stdout from an `ffmpeg -version`, as we do [in audioread](https://github.com/beetbox/audioread/blob/6a8e7283f380498d28f94a60631b65d3fa623be9/audioread/ffdec.py#L101), and then parse out the information we need, which could be somewhat brittle.

Do you know how widespread the required minimum version is? For example, it is always a useful barometer to check with version of Debian it was first included in.

Reply:
I just checked an up-to-date Debian install
```sh
$ uname -a
Linux ares 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1+deb9u3 (2019-06-16) x86_64 GNU/Linux
$ ffmpeg -version
ffmpeg version 3.2.14-1~deb9u1 Copyright (c) 2000-2019 the FFmpeg developers
built with gcc 6.3.0 (Debian 6.3.0-18+deb9u1) 20170516
configuration: --prefix=/usr --extra-version='1~deb9u1' --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --enable-gpl --disable-stripping --enable-avresample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libebur128 --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libmp3lame --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-omx --enable-openal --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libopencv --enable-libx264 --enable-shared
libavutil      55. 34.101 / 55. 34.101
libavcodec     57. 64.101 / 57. 64.101
libavformat    57. 56.101 / 57. 56.101
libavdevice    57.  1.100 / 57.  1.100
libavfilter     6. 65.100 /  6. 65.100
libavresample   3.  1.  0 /  3.  1.  0
libswscale      4.  2.100 /  4.  2.100
libswresample   2.  3.100 /  2.  3.100
libpostproc    54.  1.100 / 54.  1.100
```

It's libavfilter version is older than the included ebur128 library, but it was configured to use an external libebur128 library (`--enable-libebur128`). I only inspected the output visually, but it seems to be compatible.

To avoid a warning on compatible Debian systems we should check both the libavfilter version as well as the compile-time-configuration option.

Reply:
This check is now implemented. Though it raises an Error, not a Warning  Mostly as I did not find an existing suitable Warning class. But since we check the `--enable-libebur128` option, too, I guess that is OK?